### ISSUE TITLE
Feature: Dynamic `AnnotationsNameSpace` registery for `AnnotationNamesProvider`

### ISSUE THREAD

**Issue by emrygun** [javers/javers#1356]
## Problem
Developers may require annotations to define types and achieve required domain configurations.
https://javers.org/documentation/domain-configuration/
Current JaVers has only pre-defined annotation namespaces in it's core package:
* `JaversAnnotationsNameSpace` with static final annotation objects.
* `JPAAnnotationsNameSpace` extending `AnnotationsNamesSpace` which is registered statically into `AnnotationNamesProivder`.

So the current API forces developers to put the whole javers-core or JPA modules into their artifacts whenever they want to use these annotation type definitions.

## Solution Proposal
A dynamic registry and on `AnnotationNamesProvider` and an SPI to register (or builder method) would led developers to define their own `AnnotationsNameSpace` implementations. So developers can use their own annotations on their modules without being dependent to any other modules.

## Considerations
Creating annotations same with JPA annotation names is not an eligible solution for the problem case.

Code snippet from `JPAAnnotationsNameSpace`:
```
class JPAAnnotationsNameSpace implements AnnotationsNamespace {
    ...

    @Override
    public Set<String> getTransientPropertyAliases() {
        return Sets.asSet("Transient");
    }

    ...
}
```

**Comment by bartoszwalacik** [comment]
Hi @emrygun. Why depending on javers-core is problematic? I mean what's the scenario where you annotate your classes for Javers but you don't want to use Javers ?

**Comment by emrygun** [comment]
Hi @bartoszwalacik. Thanks for your respond :). Let me clarify.

## Scenario 1:
Consider developing an application platform contains multiple packages such as:
* model
* client
* server

And each of modules packaged on separate JARs.
So, the `model` JAR is a package including only domain models and the `client` and `server` packages are dependent on it. They use same domain models, entities, value objects and value types.

The only JAR using Javers on this platform is `server` package. It does some persistence operations and auditing them. 
The `client` has nothing to do with Javers so naturally do not want to include the javers-core.
> PS: We are awared of javers-core is a lightweight package and we'll not lose anything even if we put it into all of the packages.

So we want to declare the application domain and make fields transient on `model` package without depending on javers-core.

## Scenario 2:
Considering scenario 1, we are using some tools to create, declare domain models and they have their own conventions and annotations. 

For example:
 [immutables](https://immutables.github.io/) has it's own strategy to make class properties transient or not on their implementations. Any other tool/library may have it's own conventions and methodologies to define domains and by additional namespace providers, it'd be a lot easy to adapt them into Javers on application wise.

---

It's your decision to whether it's required or not since it's a feature suggestion. In my case, I solved this situation by some classpath scanning tricks and custom generic gson serializer/deserializers.

**Comment by bartoszwalacik** [comment]
Hi @alien11689 , first, imo AnnotationNamesProvider is used where it should be, did you see the places where it's bypassed? Second, AnnotationNamesProvider already has the extention point - it intakes List<AnnotationsNameSpace>. So what you need is a possibility to provide your custom AnnotationsNameSpace in the JaversBuilder and Spring boot configs, right? This kind of change seems easy to implement. Not sure about these processors. Please describe a use case for processors (other than your custom impl of AnnotationsNameSpace.

**Comment by alien11689** [comment]
Hi @bartoszwalacik 
I don't see or fully understand the current extenstion points:
- [AnnotationNameProvider](https://github.com/javers/javers/blob/v7.9.0/javers-core/src/main/java/org/javers/core/metamodel/scanner/AnnotationNamesProvider.java) is in package scope and the list of providers is immutable containing only `JPAAnnotationsNameSpace`.
- [JPAAnnotationsNameSpace](https://github.com/javers/javers/blob/master/javers-core/src/main/java/org/javers/core/metamodel/scanner/JPAAnnotationsNameSpace.java#L11) and [AnnotationsNameSpace](https://github.com/javers/javers/blob/master/javers-core/src/main/java/org/javers/core/metamodel/scanner/AnnotationsNameSpace.java#L8) are also package scoped
- [JaversBuilder](https://github.com/javers/javers/blob/v7.9.0/javers-core/src/main/java/org/javers/core/JaversBuilder.java) does not seem to have any method which seems to provide such option
- there are places where javers annotations are used directly without abstraction e.g. [`DiffIgnore` usage](https://github.com/search?q=repo%3Ajavers%2Fjavers+DiffIgnore+-path%3A%2F%5C%2Ftest%5C%2F%2F&type=code) or [`TypeName` usage](https://github.com/javers/javers/blob/v7.9.0/javers-core/src/main/java/org/javers/core/JaversBuilder.java#L372)
- the limitation that could be solved via the processors instead of annotations classes would solve the inconvinience where the annotation is expected to have the `value` and nothing else e.g. [here](https://github.com/javers/javers/blob/v7.9.0/javers-core/src/main/java/org/javers/core/metamodel/scanner/AnnotationNamesProvider.java#L65)
- my proposed solution would hide the annotations details in e.g. `JaversAnnotationsNameSpace` (or any other class), here detect the supported annotations (by javers or custom) and query via methods e.g. `Set<Class> getPropertyNamesAnnotations()` and then `Optional<String> getPropertyName(Annotation..., AnnotatedElement...)` when the annotation is used

**Comment by bartoszwalacik** [comment]
Of course, it's not a ready to use extention point, but it's easy to implement it as the existing design somehow anticipates it.

- change [AnnotationsNameSpace](https://github.com/javers/javers/blob/master/javers-core/src/main/java/org/javers/core/metamodel/scanner/AnnotationsNameSpace.java#L8) to public
- implement `javersBuilder.registerAnnotationsNameSpace(...)`
- trivial changes in [AnnotationNameProvider](https://github.com/javers/javers/blob/v7.9.0/javers-core/src/main/java/org/javers/core/metamodel/scanner/AnnotationNamesProvider.java)
- fix the cases you mentioned where javers uses annotations statically

For me it seems like a logical move. Still not sure how processors can fit here. Do you wan't to have also a different param name then the default "value"? So, sth like this:

```
@CustomDiffIgnoreProperties(customValueName=["field1", "field2", "getField1", "getField2"])
```

**Comment by alien11689** [comment]
I see. I thought that the extention is ready - but indeed after these steps it would be simplified.
Additionally the [`AnnotationsNameSpace`](https://github.com/javers/javers/blob/master/javers-core/src/main/java/org/javers/core/metamodel/scanner/AnnotationsNameSpace.java#L8) should cover all the annotations type used by javers.

My biggest concern here is that the annotations are not classes, not even the FQN names but the simple names in [`AnnotationNamesProvider`](https://github.com/javers/javers/blob/master/javers-core/src/main/java/org/javers/core/metamodel/scanner/AnnotationNamesProvider.java#L37). It's inconvienient when you want to exclude sth but I can agree that this is the edge case.

Processors are sth I battle tested to achieve the most from the type system and gain the extensability.
There is e.g. `TypeName` annotation that needs `value` and it needs to be `value` whereas I have own annotation for marking classes I want to store history - if I want to use my annotation as a name provider the name needs to land in the `value`. Still it will work, but is inconvienient.

I am not forcing this solution - I just put it on the table.

**Comment by alien11689** [comment]
I understand, nevertheless this is a quite easy way to add extension point to the library.

What you proposed would definietely improve also my library approach. Meantime I am testing an approach with the programatic registration of the types based on own annotations. It looks promising, but I am missing now the place in `registerType` to define renamed property. Since renaming property is currently only annotation driven then extending `PROPERTY_NAME_ANN` could be the simplest solution...
